File: backend/src/services/product.service.ts
import { ProductRepository } from '../repositories/product.repository';

export interface Product {
  id: string;
  barcode: string;
  name: string;
  price: number;
  weight?: number;
  image?: string;
  category?: string;
  createdAt: string;

  // Optional fields kept for legacy frontend compatibility
  sku?: string;
  cost?: number;
  stock?: number;
  supplierId?: string;
}

export class ProductService {
  /**
   * Fetch all products.
   */
  static async getAll(): Promise<Product[]> {
    return ProductRepository.findAll();
  }

  /**
   * Fetch product by ID.
   */
  static async getById(id: string): Promise<Product | null> {
    return ProductRepository.findById(id);
  }

  /**
   * Fetch product by barcode.
   */
  static async getByBarcode(barcode: string): Promise<Product | null> {
    return ProductRepository.findByBarcode(barcode);
  }

  /**
   * Insert product.
   */
  static async create(data: Partial<Product>): Promise<Product> {
    return ProductRepository.create(data);
  }

  /**
   * Update product details.
   */
  static async update(id: string, data: Partial<Product>): Promise<Product | null> {
    return ProductRepository.update(id, data);
  }

  /**
   * Remove a product.
   */
  static async delete(id: string): Promise<boolean> {
    return ProductRepository.delete(id);
  }
}

File: backend/app/services/product_service.py
from ..repositories.product_repository import ProductRepository
from fastapi import HTTPException

class ProductService:
    @staticmethod
    def get_all_products() -> list:
        """
        List all products in the database.
        """
        return ProductRepository.find_all()

    @staticmethod
    def get_product_by_id(product_id: str) -> dict:
        """
        Fetch a product by ID.
        """
        product = ProductRepository.find_by_id(product_id)
        if not product:
            raise HTTPException(status_code=404, detail="Product not found")
        return product

    @staticmethod
    def get_product_by_barcode(barcode: str) -> dict:
        """
        Fetch a product by barcode (used by barcode-scanners).
        """
        product = ProductRepository.find_by_barcode(barcode)
        if not product:
            raise HTTPException(status_code=404, detail=f"Product with barcode {barcode} not found")
        return product

    @staticmethod
    def create_product(data: dict) -> dict:
        """
        Register a new product barcode.
        """
        existing = ProductRepository.find_by_barcode(data["barcode"])
        if existing:
            raise HTTPException(status_code=400, detail=f"Product with barcode {data['barcode']} already exists")
        return ProductRepository.create(data)

    @staticmethod
    def update_product(product_id: str, data: dict) -> dict:
        """
        Update product info.
        """
        product = ProductRepository.update(product_id, data)
        if not product:
            raise HTTPException(status_code=404, detail="Product not found")
        return product

    @staticmethod
    def delete_product(product_id: str) -> bool:
        """
        Delete a product.
        """
        deleted = ProductRepository.delete(product_id)
        if not deleted:
            raise HTTPException(status_code=404, detail="Product not found")
        return True
